### PR TITLE
Alert rule activations integration

### PR DESCRIPTION
**Pull request by nhsiehgit**
Activated Alert Rules are conditionally subscribed to snuba queries once their activation condition has triggered (eg. on Release or Deploy deploy)

In order to persist activation run states, we've implemented the AlertRuleActivations table

On subscription for Activated Alerts Rules, we will now also initialize an AlertRuleActivation. We also update the activation witht he appropriate metric value and fill in the 'finished_at' value on expiration

this PR ensures that we are
- Creating 'activations' on conditional trigger
- Updating 'activations' on ingest
- Cleaning up and finalizing 'activations' on expiry

### CODE REVIEW

**Comment by nhsiehgit** [comment]
resolves https://github.com/getsentry/team-core-product-foundations/issues/206

File: src/sentry/incidents/models/alert_rule.py [hunk 1]
```diff
@@ -336,14 +342,24 @@ def subscribe_projects(
         )
         # NOTE: AlertRuleMonitorType.ACTIVATED will be conditionally subscribed given activation triggers
         # On activated subscription, additional query parameters will be added to the constructed query in Snuba
+        created_subscriptions = []
         if self.monitor_type == monitor_type.value:
```
Comment (ceorourke) [inline src/sentry/incidents/models/alert_rule.py:346]: Forgive me who has little context on this, so if the monitor type passed in is continuous, would there be subscriptions? And then if it's activated, we create activations for the subscriptions?

File: src/sentry/incidents/models/alert_rule.py [hunk 2]
```diff
@@ -585,14 +601,37 @@ class Meta:
...
+def update_alert_activations(
+    subscription: QuerySubscription, alert_rule: AlertRule, value: float
+) -> bool:
+    activations = alert_rule.activations.filter(
+        finished_at=None, query_subscription_id=subscription.id
+    )
+    for activation in activations:
+        activation.update(metric_value=value)
+
+    return True
```
Comment (ceorourke) [inline src/sentry/incidents/models/alert_rule.py:613]: why does this return a boolean?

File: src/sentry/incidents/models/alert_rule_activations.py [hunk 3]
```diff
@@ -63,7 +65,9 @@ class AlertRuleActivations(Model):
     # metric value represents the query results at the end of the activated time window.
     # since activated alerts are not run on a continuously shifting time window, the value
     # of this metric value will only continually tick upwards until the monitor window has expired.
-    metric_value = models.FloatField()
+    metric_value = models.FloatField(null=True)
+    # query_subscriptions are cleaned up after every run, so we keep reference to the ids rather than foreign keying
+    query_subscription_id = models.IntegerField()
```
Comment (ceorourke) [inline src/sentry/incidents/models/alert_rule_activations.py:70]: I think you need a separate migration to add this

File: src/sentry/incidents/models/alert_rule.py (hunk 1 above)
Comment (nhsiehgit) [reply src/sentry/incidents/models/alert_rule.py:346]: Yup correct

So, default is continuous.
So if the AlertRule (self) monitor_type is continuous, then we subscribe the projects.

For activated AlertRules, we'll call this subscribe_projects when the activation condition is triggered (eg. on release creation or deploy creation)

File: src/sentry/incidents/models/alert_rule.py (hunk 2 above)
Comment (nhsiehgit) [reply src/sentry/incidents/models/alert_rule.py:613]: I think this is just a success/failure indicator?
was following the existing registered callback 😅 
but i think you're right. doesn't need any return response i suppose

File: src/sentry/incidents/models/alert_rule_activations.py (hunk 3 above)
Comment (nhsiehgit) [reply src/sentry/incidents/models/alert_rule_activations.py:70]: yes ma'am you are correct

**Comment by ceorourke** [comment]
Typically migrations are in a separate PR on their own w/o code that uses the new table/column. Could you split this up?

File: src/sentry/incidents/models/alert_rule.py [hunk 4]
```diff
@@ -585,14 +601,37 @@ class Meta:
...
     if now > subscription_end:
         delete_snuba_subscription(subscription)
+        activations = alert_rule.activations.filter(
+            finished_at=None, query_subscription_id=subscription.id
+        )
+        for activation in activations:
+            activation.update(
+                metric_value=value,
+                finished_at=now,
+            )
```
Comment (markstory) [inline src/sentry/incidents/models/alert_rule.py:634]: If you don't need signals from django you can do an update on the query with `.filter().update()`

File: src/sentry/migrations/0678_add_query_sub_id_to_aar_activations.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    dependencies = [
+        ("sentry", "0677_unpickle_project_options_again"),
+    ]
+
+    operations = [
+        migrations.SeparateDatabaseAndState(
+            database_operations=[
+                migrations.RunSQL(
+                    """
+                    ALTER TABLE "sentry_alertruleactivations" ADD COLUMN "query_subscription_id" integer NOT NULL DEFAULT 0;
```
Comment (markstory) [inline src/sentry/migrations/0678_add_query_sub_id_to_aar_activations.py:30]: Shouldn't this be a nullable column? It sounds like an optional column/relation.

Comment (nhsiehgit) [reply src/sentry/migrations/0678_add_query_sub_id_to_aar_activations.py:30]: Ah
so - actually i _do_ want this to be required since activations will always be tied to a query_subscription as well (but the query subscription's can be deleted and we want this instance to persist still)

File: src/sentry/incidents/models/alert_rule.py (hunk 4 above)
Comment (nhsiehgit) [reply src/sentry/incidents/models/alert_rule.py:634]: Oh nice.
but can we iterate through and update on the filter?
Or are you saying the `.update()` will apply to all the instances returned by the filter?

**Comment by nhsiehgit** [comment]
model migration changes have been moved to [this PR](https://github.com/getsentry/sentry/pull/67483)

File: src/sentry/incidents/models/alert_rule.py (hunk 4 above)
Comment (markstory) [reply src/sentry/incidents/models/alert_rule.py:634]: >  Or are you saying the .update() will apply to all the instances returned by the filter?

Yes, all matching rows will be updated. Calling `QuerySet.update()` doesn't load any records into memory, instead it emits a query similar to `UPDATE table SET column = ? WHERE ...`

File: src/sentry/incidents/models/alert_rule.py (hunk 2 above)
Comment (saponifi3d) [reply src/sentry/incidents/models/alert_rule.py:613]: yeah, the idea was that later on we could iterate through all the responses to the registry and say if they were successfully executed or not.

File: src/sentry/migrations/0678_add_query_sub_id_to_aar_activations.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
+# Generated by Django 5.0.2 on 2024-03-21 18:46
+
+from django.db import migrations, models
+
+from sentry.new_migrations.migrations import CheckedMigration
+
+
+class Migration(CheckedMigration):
```
Comment (saponifi3d) [inline src/sentry/migrations/0678_add_query_sub_id_to_aar_activations.py:8]: is this the migration you moved to another PR or a different one? (just want to confirm this code should be in this PR)

**Review by saponifi3d — COMMENTED** [review]
lgtm once the migration is removed (if it should be?) and mark's suggestion with queryset.update is done -- django has so many foot-gun safety mechanisms 🎉

File: src/sentry/migrations/0678_add_query_sub_id_to_aar_activations.py (hunk 6 above)
Comment (nhsiehgit) [reply src/sentry/migrations/0678_add_query_sub_id_to_aar_activations.py]: Yes correct, this is the migration that was moved [here](https://github.com/getsentry/sentry/pull/67483)

**Review by saponifi3d — APPROVED** [review]

File: src/sentry/incidents/models/alert_rule.py [hunk 7]
```diff
@@ -585,13 +601,30 @@ class Meta:
...
+    subscription: QuerySubscription, alert_rule: AlertRule, value: float
+) -> bool:
     now = timezone.now()
     subscription_end = subscription.date_added + timedelta(
         seconds=subscription.snuba_query.time_window
     )
 
     if now > subscription_end:
+        alert_rule.activations.filter(finished_at=None, query_subscription=subscription).update(
+            metric_value=value, finished_at=now
```
[the remark this answers is not in the document]
Comment (nhsiehgit) [reply src/sentry/incidents/models/alert_rule.py:625]: That's a good call
I wasn't sure which trigger would fire first (i assume the update since it's registered first?)

I'll remove the metric value update from here 👍

[the remark this answers is not in the document]
Comment (nhsiehgit) [reply src/sentry/incidents/models/alert_rule.py:625]: actually 🤔 

Let me just roll both the 'update' and the 'clean expired' into a single callback
